### ISSUE TITLE
No feasible selector for ...

### ISSUE THREAD

**Issue by TheHardew** [libjxl/libjxl#895]
**Describe the bug**
Most of the images from this one camera transcoded successfully with jbrd, but a few failed:

```
$ cjxl IMG_0005.JPG - > /dev/null

JPEG XL encoder v0.7.0 cacc3da [AVX2]
Read 4608x3456 image, 144.1 MP/s
Encoding [Container | JPEG, lossless transcode, tortoise | JPEG reconstruction data | 6721-byte Exif], 16 threads.
Compressed to 4351247 bytes (2.186 bpp).
4608 x 3456, 54.37 MP/s [54.37, 54.37], 1 reps, 16 threads
./lib/jxl/fields.cc:730: JXL_FAILURE: No feasible selector for 149279
./lib/jxl/fields.cc:488: JXL_RETURN_IF_ERROR code=1: ok_
./lib/jxl/fields.cc:633: JXL_RETURN_IF_ERROR code=1: CanEncode(fields, &extension_bits, &total_bits)
./lib/jxl/jpeg/enc_jpeg_data.cc:235: JXL_RETURN_IF_ERROR code=1: Bundle::Write(jpeg_data, &writer, 0, nullptr)
Warning: failed to create JPEG reconstruction data
Including container: 4358020 bytes (2.189 bpp).
```

The error message is similar to #803, but the tail data is quite small, well below 4 MB so maybe it's a different problem?

**Expected behavior**
Ideally, jbrd would be created, but I don't know if that's actually possible.

**Screenshots**
The image: 
![IMG_0005](https://user-images.githubusercontent.com/7374061/142926568-7aa8eb70-c7d4-4afe-bce6-719dcc0bb185.JPG)

**Environment**
 - OS: Arch Linux
 - Compiler version: 13.0.0
 - CPU type: x86_64, ryzen 2700x
 - cjxl/djxl version string: cacc3da

**Comment by jonsneyers** [comment]
In this case, the problem is too many extra zero runs: spec limit is 65555, while there are 149279 of them in this jpeg.

Again, as in the case of too much tail data, `jpegtran -copy all orig.jpg >new.jpg` produces a jpeg that can be recompressed.

Again, it's a `jbrd` spec limitation that cannot be changed at this point.

**Comment by TheHardew** [comment]
Thank you, now I know I won't lose any data by stripping it.
Should I close this issue then, since the spec can't be changed?

**Comment by jonsneyers** [comment]
The spec cannot be changed, but it might still make sense to have an encode option to basically do the `jpegtran` thing for you, i.e. produce a `jbrd` that doesn't allow you to reconstruct the original jpeg file, but still allows you to reconstruct **some** jpeg file. Except for tail data (which could contain anything) and CMYK, I think all reasons why `jbrd` creation could fail due to spec limitations are things that are not really worth preserving anyway: unused quant tables, too many redundant zeroes or padding bits, etc.

I don't consider such an option to be a high-priority thing though, and not something we necessarily need to have in v1.0.

Let's keep this issue open as a low-priority feature request.

**Comment by schmorp** [comment]
Sorry to barge in, but I would urge the jpeg xl maintainers to point out these limitations clearly, because currently you can read everyhwere that jpeg-xl can losslessly transcode to jpeg and back (which is true, because it can transcode some jpeg files, but not all).

I can tell you that archivists are very happy (thats sarcasm :) when they read about features of a format, only to later find out they haven't been told the full truth - and even for normal users, it has to be kept in mind that most people trying to transcode jpeg are not responsible for the "not worthy of preservation" state of the files they transcode.

So while this cnanot be implemented for spec reasons, it should be more clearly pointed out in documentation and advertising ("many/most/the vast majority jpeg files can be losslessly transcoded").

This also has some overlap with my feature request #950, to reconstruct a "pixel-equivalent" JPEG file, even when a bit-identical reconstrruction is not possible for whatever reason.

**Comment by jonsneyers** [comment]
I agree it would be useful to document and clarify the limitations of what kind of JPEG files can be losslessly transcoded. If I recall correctly, we at some point crawled lots of JPEGs from the web and 99.9something percent of them could be losslessly transcoded, but that probably says more about web images than anything else.

The main limitations of the top of my head are:
- No CMYK, only YCbCr or RGB or grayscale
- Only 4:4:4, 4:2:0, 4:2:2 and 4:4:0 chroma subsampling (or rather, anything with at most a 2x upsampling of each channel)
- Not more than ~4 MB of tail data (bytes after the end of the jpeg bitstream)
- No redundant (unused) quantization tables that are different from the last non-redundant quantization table
- Not too many redundant zero runs
- Only the subset of jpeg that is commonly implemented in decoders, so not arithmetic coding, hierarchical jpeg, lossless jpeg, etc.
- probably some more

Some of these limitations make "DCT-coeff-equivalent" encoding impossible, others only make bit-identical reconstruction impossible.

It would be good to complete this list of limitations and add it to the documentation. It would also be good to have better error handling in the encode API so applications can see the difference between "it is a corrupt jpeg", "the jpeg is probably valid but I could not represent the jpeg image data losslessly", "I could represent the jpeg image data losslessly but bit-identical file reconstruction is not possible" and "something went wrong during encode for whatever reason".

**Comment by william-silversmith** [comment]
Hi! I'm working with large numbers of scientific grayscale images and we are considering using jpeg/jxl transcoding (in both directions) for various projects. I was experimenting with the below attached JXL image and for some reason it cannot transcode losslessly between JPEG and JXL. Is there an obvious reason as to why?

Here's the `jxlinfo` for the image:

```
JPEG XL image, 64x4096, lossy, 8-bit Grayscale
Color space: Grayscale, D65, sRGB transfer function, rendering intent: Relative
```

When running `djxl`:

```bash
$ djxl bad3.jxl bad3.jpeg
JPEG XL decoder v0.8.2 0.8.2 [NEON]
Read 586 compressed bytes.
Warning: could not decode losslessly to JPEG. Retrying with --pixels_to_jpeg...
Decoded to pixels.
64 x 4096, 44.23 MP/s [44.23, 44.23], 1 reps, 8 threads.
```

[bad3.jxl.zip](https://github.com/user-attachments/files/15829094/bad3.jxl.zip)

**Comment by jonnyawsom3** [comment]
@william-silversmith Apologies for the wait. That JXL file uses lossy encoding, and not Lossless JPEG Transcoding.
As far as I'm aware, only cjxl currently does the transcode when a JPEG file is input. Most tools like FFMPEG, ImageMagick, VIPS, ect decode to pixels before passing to libjxl, meaning it doesn't have access to the DCT for lossless compression.
